tests/helpers_tests/test_condition.py

```diff
@@ -721,3 +723,594 @@ def test_audience_condition_deserializer_defaults(self):
     self.assertIsNone(items[1])
     self.assertIsNone(items[2])
     self.assertIsNone(items[3])
+
+
+class CustomAttributeConditionEvaluatorLogging(base.BaseTest):
+
+  def setUp(self):
+    base.BaseTest.setUp(self)
+    self.mock_client_logger = mock.MagicMock()
+
+  def test_evaluate__match_type__invalid(self):
+    log_level = 'warning'
+    condition_list = [['favorite_constellation', 'Lacerta', 'custom_attribute', 'regex']]
+    user_attributes = {}
+
+    with mock.patch('optimizely.logger.reset_logger', return_value=self.mock_client_logger):
+      evaluator = condition_helper.CustomAttributeConditionEvaluator(
+        condition_list, user_attributes, self.mock_client_logger
+      )
+
+    expected_condition_log = {
+      "name": 'favorite_constellation',
+      "value": 'Lacerta',
+      "type": 'custom_attribute',
+      "match": 'regex'
+    }
+
+    self.assertIsNone(evaluator.evaluate(0))
+
+    mock_log = getattr(self.mock_client_logger, log_level)
+    mock_log.assert_called_once_with((
+      'Audience condition "{}" uses an unknown match '
+      'type. You may need to upgrade to a newer release of the Optimizely SDK.')
+        .format(json.dumps(expected_condition_log)))
```

Comment:
I'd still _prefer_ to avoid using `.format` in our assertions since it's an [ugly mirror](https://jasonrudolph.com/blog/2008/07/30/testing-anti-patterns-the-ugly-mirror/), but it does help that we're only using it for the conditions now and not for other template variables in our log messages. I guess this is okay.

Reply:
Yeah, for consistent unit tests we do need to use format in cases where representation varies across different python versions. Same for variable types.